apps/ejabberd/src/ejabberd_s2s_out.erl

```diff
@@ -467,84 +450,16 @@ wait_for_features({xmlstreamelement, El}, StateData) ->
                           Acc
                   end, {false, false, false}, Els),
-            if
-                (not SASLEXT) and (not StartTLS) and
-                StateData#state.authenticated ->
-                    send_queue(StateData, StateData#state.queue),
-                    ?INFO_MSG("Connection established: ~s -> ~s",
-                              [StateData#state.myname, StateData#state.server]),
-                    ejabberd_hooks:run(s2s_connect_hook,
-                                       [StateData#state.myname,
-                                        StateData#state.server]),
-                    {next_state, stream_established,
-                     StateData#state{queue = queue:new()}};
-                SASLEXT and StateData#state.try_auth and
-                (StateData#state.new /= false) ->
-                    send_element(StateData,
-                                  #xmlel{name = <<"auth">>,
-                                         attrs = [{<<"xmlns">>, ?NS_SASL},
-                                                  {<<"mechanism">>, <<"EXTERNAL">>}],
-                                          children = [#xmlcdata{content = jlib:encode_base64(
-                                                                            StateData#state.myname)}]}),
-                     {next_state, wait_for_auth_result,
-                      StateData#state{try_auth = false}, ?FSMTIMEOUT};
-                 StartTLS and StateData#state.tls and
-                  (not StateData#state.tls_enabled) ->
-                     send_element(StateData,
-                                  #xmlel{name = <<"starttls">>,
-                                         attrs = [{<<"xmlns">>, ?NS_TLS}]}),
-                     {next_state, wait_for_starttls_proceed, StateData,
-                      ?FSMTIMEOUT};
-                 StartTLSRequired and (not StateData#state.tls) ->
-                     ?DEBUG("restarted: ~p", [{StateData#state.myname,
-                                               StateData#state.server}]),
-                     ejabberd_socket:close(StateData#state.socket),
-                     {next_state, reopen_socket,
-                      StateData#state{socket = undefined,
-                                      use_v10 = false}, ?FSMTIMEOUT};
-                 StateData#state.db_enabled ->
-                     send_db_request(StateData);
-                 true ->
-                     ?DEBUG("restarted: ~p", [{StateData#state.myname,
-                                               StateData#state.server}]),
-                     % TODO: clear message queue
-                     ejabberd_socket:close(StateData#state.socket),
-                     {next_state, reopen_socket, StateData#state{socket = undefined,
-                                                                 use_v10 = false}, ?FSMTIMEOUT}
-            end;
+            try_establish_conn({SASLEXT, StartTLS, StartTLSRequired, StateData});
```

Comment:
Connection is already established and the results of this function may be very different. What do you think about `handle_parsed_features`?